[FEATURE][ML] Implement data frame evaluate API

dimitris-athanasiou opened the pull request:
Adds a new API:

```
POST _ml/data_frame/_evaluate
```

which enables evaluating various ML methods.

elasticmachine commented:
Pinging @elastic/ml-core

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:
```diff
@@ -842,4 +848,16 @@ static long machineMemoryFromStats(OsStats stats) {
     }
+
+    @Override
+    public List<NamedWriteableRegistry.Entry> getNamedWriteables() {
```
We have a data frame plugin but this is part of ml?

droberts195 replied:
The data frame plugin is for data frame _transforms_, and is basic licensed.  The data frame _analytics_ is in the ML plugin and is platinum licensed.

But the fact that you asked this shows that the docs that explain the distinction to users will need careful thought.

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/EvaluateDataFrameAction.java:
```diff
@@ -0,0 +1,207 @@
+public class EvaluateDataFrameAction extends Action<EvaluateDataFrameAction.Response> {
+
+    public static final EvaluateDataFrameAction INSTANCE = new EvaluateDataFrameAction();
+    public static final String NAME = "cluster:monitor/xpack/ml/data_frame/evaluate";
```
This endpoint is part of data frame _analytics_, which is part of the ML plugin, so _must_ have action names that begin with `(admin|monitor)/xpack/ml` to match the ML plugin cluster privileges.

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportEvaluateDataFrameAction.java:
```diff
@@ -0,0 +1,49 @@
+        );
+
+        client.execute(SearchAction.INSTANCE, searchRequest, ActionListener.wrap(
+            searchResponse -> evaluation.evaluate(searchResponse, resultListener),
```
I thought I had commented on this previously but the message seems to have been lost. 

`evaluation.evaluate()` should run in a different thread

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/softclassification/ScoreByThresholdResult.java:
```diff
@@ -0,0 +1,63 @@
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        for (int i = 0; i < thresholds.length; i++) {
+            builder.field(String.valueOf(thresholds[i]), scores[i]);
```
This follows the pattern used in the percentiles aggregation. I have discussed with Martijn from the clients team who rested my concern by saying the format has posed no problems to clients. I have considered the idea of having the thresholds into an array, however the downside of that would be that you have to assume there is mapping between scores and thresholds based on the array index. There is no clear winner here so I have settled for following the percentiles paradigm.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/softclassification/AucRoc.java:
```diff
@@ -0,0 +1,347 @@
+
+    private static double[] percentilesArray(Percentiles percentiles) {
+        double[] result = new double[99];
+        percentiles.iterator().forEachRemaining(percentile -> result[((int) percentile.getPercent()) - 1] = percentile.getValue());
```
the iterator object does not have `forEach` but I can just call `percentiles.forEach` so I'll do that.

dimitris-athanasiou commented:
I have updated the PR addressing review comments plus some additional changes. It seems I still have problems with the security tests. I'll look into it and update.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/MlEvaluationNamedXContentProvider.java:
```diff
@@ -0,0 +1,73 @@
+            BinarySoftClassification::new));
+
+        // Evaluation Results
+        namedWriteables.add(new NamedWriteableRegistry.Entry(EvaluationResult.class, MetricListEvaluationResult.WRITEABLE_NAME,
```
That's a good point. I was meaning to rename all others to `WRITEABLE_NAME` too. The `writeableName` and the actual name occasionally match but they are conceptually different things. For example, the results for precision and recall are the same object. That object's writeable name is `score_by_threshold_result` as that's enough to identify how to read/write it on the wire. However, its name in JSON differs depending on whether it is precision or recall. Does this make sense?

dimitris-athanasiou replied:
I have added a commit which hopefully clarifies the difference between these.

dimitris-athanasiou commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/EvaluateDataFrameAction.java:
```diff
@@ -0,0 +1,207 @@
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(result.getName(), result);
```
This allows the response to have the evaluation type at the top level.

The response for `binary_soft_classification` currently is (for example): 

```
{
    "binary_soft_classification": {
        "auc_roc": {
            "score": 0.7680318097975196
        }
    }
}
```

If I took that away it would become:

```
{
  "auc_roc": {
      "score": 0.7680318097975196
  }
}
```

The problem with that would be that a client would not know how to parse this as different evaluations would result in different responses.

dimitris-athanasiou commented:
@davidkyle I tried writing unit tests for those. However, there is no way to create the aggregation results but to mock them. It then results in lots of mocking just to test the single line where the actual computation happens. Instead, I have added YAML tests that test those.

dimitris-athanasiou commented:
@davidkyle I added unit tests for the evaluate method of precision, recall and confusion_matrix.

dimitris-athanasiou commented:
run elasticsearch-ci/packaging-sample
